os/binfmt/binfmt_execmodule.c

```diff
@@ -0,0 +1,309 @@
+ *   failure.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_BINFMT_CONSTRUCTORS
+static void exec_ctors(FAR void *arg)
+{
+	FAR const struct binary_s *binp = (FAR const struct binary_s *)arg;
+	binfmt_ctor_t *ctor = binp->ctors;
+	int i;
+
+	/* Execute each constructor */
+
+	for (i = 0; i < binp->nctors; i++) {
+		binfo("Calling ctor %d at %p\n", i, (FAR void *)ctor);
+
+		(*ctor)();
+		ctor++;
+	}
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: exec_module
+ *
+ * Description:
+ *   Execute a module that has been loaded into memory by load_module().
+ *
+ * Returned Value:
+ *   This is a internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int exec_module(FAR const struct binary_s *binp)
+{
+	FAR struct task_tcb_s *tcb;
+#if defined(CONFIG_ARCH_ADDRENV) && defined(CONFIG_BUILD_KERNEL)
+	save_addrenv_t oldenv;
+#endif
+	FAR uint32_t *stack;
+	pid_t pid;
+	int ret;
+
+	/* Sanity checking */
+
+#ifdef CONFIG_DEBUG
+	if (!binp || !binp->entrypt || binp->stacksize <= 0) {
+		return -EINVAL;
+	}
+#endif
+
+	binfo("Executing %s\n", binp->filename);
+
+	/* Allocate a TCB for the new task. */
+
+	tcb = (FAR struct task_tcb_s *)kmm_zalloc(sizeof(struct task_tcb_s));
+	if (!tcb) {
+		return -ENOMEM;
+	}
+#if defined(CONFIG_ARCH_ADDRENV) && defined(CONFIG_BUILD_KERNEL)
+	/* Instantiate the address environment containing the user heap */
+
+	ret = up_addrenv_select(&binp->addrenv, &oldenv);
+	if (ret < 0) {
+		berr("ERROR: up_addrenv_select() failed: %d\n", ret);
+		goto errout_with_tcb;
+	}
+#endif
+
+	/* Allocate the stack for the new task.
+	 *
+	 * REVISIT:  This allocation is currently always from the user heap.  That
+	 * will need to change if/when we want to support dynamic stack allocation.
+	 */
+
+	stack = (FAR uint32_t *)kumm_malloc(binp->stacksize);
+	if (!stack) {
+		ret = -ENOMEM;
+		goto errout_with_addrenv;
+	}
+
+	/* Initialize the task */
+
+	ret = task_init((FAR struct tcb_s *)tcb, binp->filename, binp->priority, stack, binp->stacksize, binp->entrypt, binp->argv);
+	if (ret < 0) {
+		ret = -get_errno();
+		berr("task_init() failed: %d\n", ret);
+		goto errout_with_addrenv;
+	}
+
+	/* We can free the argument buffer now.
+	 * REVISIT:  It is good to free up memory as soon as possible, but
+	 * unfortunately here 'binp' is 'const'.  So to do this properly, we will
+	 * have to make some more extensive changes.
+	 */
+
+	binfmt_freeargv((FAR struct binary_s *)binp);
+
+	/* Note that tcb->flags are not modified.  0=normal task */
+	/* tcb->flags |= TCB_FLAG_TTYPE_TASK; */
```

Comment:
Can we remove the line if it is not necessary instead of commenting?

Reply:
It is useful, no need to remove